### PR TITLE
Add 'gsctl update organization set-credentials' command

### PR DESCRIPTION
**Pull request by marians**
For https://github.com/giantswarm/giantswarm/issues/3987

### Preview

```
$ gsctl update org sc --help
Set the credentials used to create and operate the clusters of an organization.

Setting credentials of an organization will result in all future cluster
being run in the account/subscription referenced by the credentials. Once
credentials are set for an organization, this cannot be undone.

For details on how to prepare the account/subscription, consult the documentation at

  - https://docs.giantswarm.io/guides/prepare-aws-account-for-tenant-clusters/ (AWS)
  - https://docs.giantswarm.io/guides/prepare-azure-subscription-for-tenant-clusters/ (Azure)

Usage:
  gsctl update organization set-credentials [flags]

Aliases:
  set-credentials, sc

Examples:

  gsctl update organization set-credentials -o acme \
    --aws-operator-role arn:aws:iam::<AWS-ACCOUNT-ID>:role/GiantSwarmAWSOperator \
    --aws-admin-role arn:aws:iam::<AWS-ACCOUNT-ID>:role/GiantSwarmAdmin

  gsctl update organization set-credentials -o acme \
    --azure-subscription-id <AZURE-SUBSCRIPTION-ID> \
    --azure-tenant-id <AZURE-TENANT-ID> \
    --azure-client-id <AZURE-CLIENT-ID> \
    --azure-secret-key <AZURE-SECRET-KEY>


Flags:
      --aws-admin-role string          AWS ARN of the role to be used by Giant Swarm staff
      --aws-operator-role string       AWS ARN of the role to use for operating clusters
      --azure-client-id string         ID of the Azure service principal to use for operating clusters
      --azure-secret-key string        Secret key for the Azure service principal to use for operating clusters
      --azure-subscription-id string   ID of the Azure subscription to run clusters in
      --azure-tenant-id string         ID of the Azure tenant to run clusters in
  -h, --help                           help for set-credentials
  -o, --organization string            ID of the organization to set credentials for

Global Flags:
      --auth-token string   Authorization token to use
      --config-dir string   Configuration directory path to use (default "/Users/marian/.config/gsctl")
  -e, --endpoint string     The API endpoint to use
  -v, --verbose             Print more information
```

```
./gsctl update org sc -o giantswarm \
  --aws-operator-role arn:aws:iam::012345678:role/GiantSwarmAWSOperator \
  --aws-admin-role arn:aws:iam::012345678:role/GiantSwarmAdmin
Credentials set successfully
The credentials are stored with the unique ID 'a354sh'.
```

### CODE REVIEW

File: commands/update_organization_setcredentials.go [hunk 1]
```diff
@@ -0,0 +1,351 @@
...
+
+	if err != nil {
+		handleCommonErrors(err)
+
+		// From here on we handle errors that can only occur in this command
+		headline := ""
+		subtext := ""
+
+		switch {
+		case err.Error() == "":
```
Comment (kopiczko) [inline commands/update_organization_setcredentials.go]: Why does it need to be checked?

File: client/clienterror/clienterror.go [hunk 2]
```diff
@@ -247,6 +247,31 @@ func New(err error) *APIError {
 		}
 	}
 
+	// add credentials
+	if addCredentialsConflict, ok := err.(*organizations.AddCredentialsConflict); ok {
+		return &APIError{
+			ErrorMessage:   "The organization has credentials already",
+			ErrorDetails:   "Credentials are immutible and an organization can only have one credential set.",
```
Comment (JosephSalisbury) [inline client/clienterror/clienterror.go]: `immutable`

File: client/clienterror/clienterror.go [hunk 3]
```diff
@@ -247,6 +247,31 @@ func New(err error) *APIError {
...
+			HTTPStatusCode: http.StatusConflict,
+			OriginalError:  addCredentialsConflict,
+		}
+	}
+	if addCredentialsUnauthorized, ok := err.(*organizations.AddCredentialsUnauthorized); ok {
+		return &APIError{
+			ErrorMessage:   "Unauthorized",
+			ErrorDetails:   "You have no permission to set credentials for this organization.",
```
Comment (JosephSalisbury) [inline client/clienterror/clienterror.go]: `You do not have permission to set...` reads slightly easier for me

File: commands/commands.go [hunk 4]
```diff
@@ -53,6 +53,9 @@ var (
 	// full flag. if set, output must not be truncated.
 	cmdFull bool
 
+	// organizaiton ID as passed by the user as a flag
+	cmdOrganization string
```
Comment (JosephSalisbury) [inline commands/commands.go]: maybe `cmdOrganizationID` then?

File: commands/commands.go [hunk 5]
```diff
@@ -53,6 +53,9 @@ var (
 	// full flag. if set, output must not be truncated.
 	cmdFull bool
 
+	// organizaiton ID as passed by the user as a flag
```
Comment (JosephSalisbury) [inline commands/commands.go]: `organization`

File: commands/update_organization_setcredentials.go [hunk 6]
```diff
@@ -0,0 +1,351 @@
...
+var (
+
+	// UpdateOrgSetCredentialsCommand performs the "update organizatio set-credentials" function
+	UpdateOrgSetCredentialsCommand = &cobra.Command{
+		Use:     "set-credentials",
+		Aliases: []string{"sc"},
+		Short:   "Set credentials of an organization",
+		Long: `Set the credentials used to create and operate the clusters of an organization.
+
+Setting credentials of an organization will result in all future cluster
```
Comment (JosephSalisbury) [inline commands/update_organization_setcredentials.go]: `future clusters`

File: commands/update_organization_setcredentials.go [hunk 7]
```diff
@@ -0,0 +1,351 @@
...
+being run in the account/subscription referenced by the credentials. Once
+credentials are set for an organization, this cannot be undone.
```
Comment (JosephSalisbury) [inline commands/update_organization_setcredentials.go]: `this currently cannot be undone` sounds better, we probably will support this eventually

File: commands/update_organization_setcredentials.go [hunk 8]
```diff
@@ -0,0 +1,351 @@
...
+
+  gsctl update organization set-credentials -o acme \
+    --azure-subscription-id <AZURE-SUBSCRIPTION-ID> \
+    --azure-tenant-id <AZURE-TENANT-ID> \
+    --azure-client-id <AZURE-CLIENT-ID> \
+    --azure-secret-key <AZURE-SECRET-KEY>
+`,
+
+		// PreRun checks a few general things, like authentication and flags
+		// plausibility.
```
Comment (JosephSalisbury) [inline commands/update_organization_setcredentials.go]: `plausibility` isn't the right term here. `compatibility` is probably more correct

File: commands/update_organization_setcredentials.go [hunk 9]
```diff
@@ -0,0 +1,351 @@
...
+		PreRun: updateOrgSetCredentialsPreRunOutput,
+
+		// Run calls the business function and prints results and errors.
+		Run: updateOrgSetCredentialsRunOutput,
+	}
+
+	// AWS operator role ARN as passed by the user via flags
+	cmdAWSOperatorRoleARN string
```
Comment (JosephSalisbury) [inline commands/update_organization_setcredentials.go]: In the same way as we have the azure flags in one block, we can probably put the aws ones in one block. i had to think here

File: commands/update_organization_setcredentials.go [hunk 10]
```diff
@@ -0,0 +1,351 @@
...
+		subtext = "Setting credentials is only supported on AWS and Azure installations."
+	case IsRequiredFlagMissingError(err):
+		headline = "Missing flag: " + err.Error()
+		subtext = "Please use --help to see details regarding the command's usage."
+	case IsConflictingFlagsError(err):
+		headline = "Conflicting flags"
+		subtext = "Please use only AWS or Azure related flags with this installation. See --help for details."
+	case IsOrganizationNotFoundError(err):
+		headline = fmt.Sprintf("Organization '%s' not found", args.organizationID)
+		subtext = "The specified organization does not exist, or you are not a member. Please check the exact upper-/lower case spelling."
```
Comment (JosephSalisbury) [inline commands/update_organization_setcredentials.go]: `upper/lower`

**Comment by marians** [comment]
PTAL @JosephSalisbury

File: commands/update_organization_setcredentials.go [hunk 11]
```diff
@@ -0,0 +1,349 @@
...
+	}
+
+	handleCommonErrors(err)
+
+	// From here on we handle errors that can only occur in this command
+	headline := ""
+	subtext := ""
+
+	switch {
+	case err.Error() == "":
```
Comment (TheoBrigitte) [inline commands/update_organization_setcredentials.go]: What is this case for ?
Isnt' this already catched by `err == nil` ?

File: commands/update_organization_setcredentials.go [hunk 12]
```diff
@@ -0,0 +1,349 @@
...
+			}
+		}
+		return nil, microerror.Mask(err)
+	}
+
+	// Location header returned is in the format
+	// /v4/organizations/myorg/credentials/{credential_id}/
+	segments := strings.Split(response.Location, "/")
+	result := &updateOrgSetCredentialsResult{
+		credentialID: segments[len(segments)-2],
```
Comment (TheoBrigitte) [inline commands/update_organization_setcredentials.go]: What is this magical **2** here for ?

File: commands/update.go [hunk 13]
```diff
@@ -0,0 +1,16 @@
+package commands
+
+import "github.com/spf13/cobra"
+
+var (
+	// UpdateCommand is the command to modify resources
+	UpdateCommand = &cobra.Command{
+		Use:   "update",
+		Short: "Modify organization details",
+		Long:  `Modify details of an organization`,
```
[the remark this answers is not in the document]
Comment (marians) [reply commands/update.go]: I don't understand that

File: commands/update_organization_setcredentials.go (hunk 12 above)
Comment (marians) [reply commands/update_organization_setcredentials.go]: It's to return the second last item from the slice. See https://play.golang.org/p/krFhD0fU_K-

File: commands/update_organization_setcredentials.go [hunk 14]
```diff
@@ -0,0 +1,349 @@
...
+	"github.com/spf13/cobra"
+)
+
+const (
+	updateOrgSetCredentialsActivityName = "set-org-credentials"
+)
+
+var (
+
+	// UpdateOrgSetCredentialsCommand performs the "update organizatio set-credentials" function
```
[the remark this answers is not in the document]
Comment (marians) [reply commands/update_organization_setcredentials.go]: fixed

File: commands/update_organization_setcredentials.go (hunk 11 above)
Comment (marians) [reply commands/update_organization_setcredentials.go]: Removed

File: commands/update_organization_setcredentials.go (hunk 1 above)
Comment (marians) [reply commands/update_organization_setcredentials.go]: Removed

**Pull request merged**